yaf-src/airframe/doc/filedaemon.1
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "FILEDAEMON 1"
.TH FILEDAEMON 1 "20-Dec-2023" "2.15.0" "airframe"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
filedaemon \- Invoke another program on files matching a glob pattern
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 9
\&    filedaemon  \-\-in INPUT_GLOB [\-\-out OUTPUT_DIR]
\&                \-\-nextdir PROCESSED_INPUT_DIRECTORY
\&                \-\-faildir FAILED_INPUT_DIRECTORY
\&                [\-\-extension OUTPUT_EXTENSION]
\&                [\-\-poll POLLING_DELAY] [\-\-lock]
\&                [\-\-pidfile PID_FILE] [\-\-no\-daemon]
\&                [\-\-log LOG_SPECIFIER] [\-\-loglevel LOG_LEVEL]
\&                [\-\-verbose] [\-\-version]
\&                \-\- PROGRAM [PROGRAM_ARGS]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBfiledaemon\fR operates on input files matching a \fBglob\fR\|(3) pattern.  Each matching
input file is passed to a child program (specified in the \s-1PROGRAM\s0 argument)
which is expected to read input data from standard input and write the results
to standard output.  \fBfiledaemon\fR's job is simply to handle the mechanics of
directory polling, file globbing, and routing of input and output files on
behalf of the child program.
.PP
Use of two dashes (\-\-) after all filedaemon command-line switches allows
\&\s-1PROGRAM_ARGS\s0 to be interpreted by the \s-1PROGRAM\s0 rather than \fBfiledaemon\fR itself.
While they are not strictly required if you do not need to pass arguments to
\&\s-1PROGRAM,\s0 they should be used for consistency.
.SH "OPTIONS"
.IX Header "OPTIONS"
.SS "I/O Options"
.IX Subsection "I/O Options"
These options control the file locations for input and output files.
.IP "\fB\-\-in\fR \fI\s-1INPUT_GLOB\s0\fR" 4
.IX Item "--in INPUT_GLOB"
\&\fI\s-1INPUT_GLOB\s0\fR is a file glob pattern, which must be  escaped or quoted to
prevent the shell expansion.  Files that match this pattern will be processed
by filedaemon.  This option is required.
.IP "\fB\-\-out\fR \fI\s-1OUTPUT_DIR\s0\fR" 4
.IX Item "--out OUTPUT_DIR"
\&\fI\s-1OUTPUT_DIR\s0\fR is a directory in which to place output files.  The directory
must exist prior to invoking \fBfiledaemon\fR, and any files in the directory
that match the names of output files will be overwritten.  If not specified,
the current working directory will be used.
.SS "Daemon Options"
.IX Subsection "Daemon Options"
These options control what is done with processed input files, file locking,
and other options to facilitate operation as a file daemon.
.IP "\fB\-\-nextdir\fR \fI\s-1PROCESSED_INPUT_DIRECTORY\s0\fR" 4
.IX Item "--nextdir PROCESSED_INPUT_DIRECTORY"
When reading from files, if this option is present, input files will be
moved to \fI\s-1PROCESSED_INPUT_DIRECTORY\s0\fR after they are successfully processed.
The special string \fBdelete\fR will cause successfully processed input to be
removed instead. This option is required.
.IP "\fB\-\-faildir\fR \fI\s-1FAILED_INPUT_DIRECTORY\s0\fR" 4
.IX Item "--faildir FAILED_INPUT_DIRECTORY"
When reading from files, if this option is present, input files will be
moved to \fI\s-1FAILED_INPUT_DIRECTORY\s0\fR if processing failed.
The special string \fBdelete\fR will cause failed input to be
removed instead. This option is required.
.IP "\fB\-\-extension\fR \fI\s-1OUTPUT_EXTENSION\s0\fR" 4
.IX Item "--extension OUTPUT_EXTENSION"
Replace the input file's extension with \fI\s-1OUTPUT_EXTENSION\s0\fR.  For
example, if an input file is named \*(L"foo.txt\*(R", and this option is \*(L"out\*(R",
then the output file will be named \*(L"foo.out\*(R".  If the input file has no
extension, then this option's value will be appended to the filename.
If this option is not specified, output files will have the same name

yaf-src/NEWS
Version 2.15.0: 2023-12-21
==========================

Enhanced the deep packet inspection capabilities for SSH connections to include negotiated algorithms and HASSH hash.

Added the JA3 hash to the DPI for TLS connections.

Added support for reading VxLAN-encapsulated packets, Geneve-encapsulated packets, and Geneve-encapsulated VxLAN-encapsulated packets.

Fixed TLS certificate parsing to be more selective on which values are stored in the list of sslObjectType-sslObjectValue pairs.

Fixed a potential bug in the Shannon entropy calculation that may cause small differences in calculated values.

Version 3.0.0.alpha4: 2023-12-21
================================

Added the ability to decode VxLAN-encapsulated packets, Geneve-encapsulated packets, and Geneve-encapsulated VxLAN-encapsulated packets.

Added a --no-mpls option to disable export of MPLS labels when YAF has been built with MPLS support.

Fixed TLS certificate parsing to be more selective on which values are stored in the list of sslObjectType-sslObjectValue pairs.

Version 3.0.0.alpha3: 2023-07-18
================================

Changed DNS deep packet inspection to produce names and text records with escape codes for special characters (non-ASCII, non-printable, special whitespace, and label-internal dots in names).

Made DNS deep packet inspection more strict about parsing malformed DNS Resource Records across RR boundaries within the packet.

Enhanced the --time and --etime options of yafMeta2Pcap to accept a human-readable timestamp in addition to milliseconds.

Changed the destination of --version output to the standard output.

Changed yaf to only export the fingerprint-related elements (firstPacketBanner, etc) when the --fpexport option is given.  (Requires YAF to be built with --enable-fpexporter.)

Changed yaf to only export the p0f-related elements (osName, etc) when the --p0fprint option is given.  (Requires YAF to be built with --with-p0f.)

Fixed a crash in YAF that occurs when it is built with GLib 2.75.3 or newer.

Version 2.14.0: 2023-03-23
==========================

Changed DNS deep packet inspection to produce names and text records with escape codes for special characters (non-ASCII, non-printable, special whitespace, and label-internal dots in names).

Made DNS deep packet inspection more strict about parsing malformed DNS Resource Records across RR boundaries within the packet.

Changed destination of --version output to the standard output.

Fixed a crash in YAF that occurs when it is built with GLib 2.75.3 or newer.

Version 3.0.0.alpha2: 2023-02-09
================================

Enhanced the deep packet inspection capabilities for SSH connections to include negotiated algorithms and HASSH hash.

Added the JA3 hash to the DPI for TLS connections.

Made several changes to the yafDPIRules.conf file for applabels written as C plugins: Allow the user to disable the export of arbitrary DPI elements and SMTP headers. Allow a protocol to be specified. Moved the regex definitions from C to yafDPIRules.conf.

Increased the maximum payload that YAF may capture for performing DPI.

Fixed a potential bug in the Shannon entropy calculation that may cause small differences in calculated values.

Version 2.13.0: 2023-02-09
==========================

Added ability for yaf to limit payload export to a named set of applabels.

Increased the maximum payload that YAF may capture for performing DPI.

Added support for recent releases of nDPI.

Added yaf.init to the list of installed files.

Stopped export of full flow template that is never used for data records.

Fixed minor bug in --version where Compact IPv4 support always reported NO.

Fixed bugs in regular expressions for nntpResponseRegex and smtpURLRegex.

Version 3.0.0.alpha1: 2022-02-28
================================


yaf-src/airframe/src/filedaemon.c
/*
 *  Copyright 2007-2023 Carnegie Mellon University
 *  See license information in LICENSE.txt.
 */
/*
 *  filedaemon.c
 *  Invokes yaf, naf, etc. as a file daemon process
 *
 *  ------------------------------------------------------------------------
 *  Authors: Tony Cebzanov
 *  ------------------------------------------------------------------------
 *  @DISTRIBUTION_STATEMENT_BEGIN@
 *  YAF 2.15.0
 *
 *  Copyright 2023 Carnegie Mellon University.
 *
 *  NO WARRANTY. THIS CARNEGIE MELLON UNIVERSITY AND SOFTWARE ENGINEERING
 *  INSTITUTE MATERIAL IS FURNISHED ON AN "AS-IS" BASIS. CARNEGIE MELLON
 *  UNIVERSITY MAKES NO WARRANTIES OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
 *  AS TO ANY MATTER INCLUDING, BUT NOT LIMITED TO, WARRANTY OF FITNESS FOR
 *  PURPOSE OR MERCHANTABILITY, EXCLUSIVITY, OR RESULTS OBTAINED FROM USE OF
 *  THE MATERIAL. CARNEGIE MELLON UNIVERSITY DOES NOT MAKE ANY WARRANTY OF
 *  ANY KIND WITH RESPECT TO FREEDOM FROM PATENT, TRADEMARK, OR COPYRIGHT
 *  INFRINGEMENT.
 *
 *  Licensed under a GNU GPL 2.0-style license, please see LICENSE.txt or
 *  contact permission@sei.cmu.edu for full terms.
 *
 *  [DISTRIBUTION STATEMENT A] This material has been approved for public
 *  release and unlimited distribution.  Please see Copyright notice for
 *  non-US Government use and distribution.
 *
 *  GOVERNMENT PURPOSE RIGHTS - Software and Software Documentation
 *  Contract No.: FA8702-15-D-0002
 *  Contractor Name: Carnegie Mellon University
 *  Contractor Address: 4500 Fifth Avenue, Pittsburgh, PA 15213
 *
 *  The Government's rights to use, modify, reproduce, release, perform,
 *  display, or disclose this software are restricted by paragraph (b)(2) of
 *  the Rights in Noncommercial Computer Software and Noncommercial Computer
 *  Software Documentation clause contained in the above identified
 *  contract. No restrictions apply after the expiration date shown
 *  above. Any reproduction of the software or portions thereof marked with
 *  this legend must also reproduce the markings.
 *
 *  This Software includes and/or makes use of Third-Party Software each
 *  subject to its own license.
 *
 *  DM23-2313
 *  @DISTRIBUTION_STATEMENT_END@
 *  ------------------------------------------------------------------------
 */

#define _AIRFRAME_SOURCE_
#include <airframe/autoinc.h>
#include <airframe/airopt.h>
#include <airframe/logconfig.h>

static char          *fd_inspec          = NULL;
static char          *fd_outspec         = NULL;
static char          *fd_nextdir         = NULL;
static char          *fd_faildir         = NULL;
static char          *fd_outext          = NULL;
static uint32_t       fd_poll_delay      = 30;
static gboolean       fd_lock            = FALSE;
static gsize          fd_bufsize         = 1024;
static char          *fd_pidfile         = NULL;
static pid_t          fd_pid             = 0;
static gboolean       fd_nodaemon        = FALSE;

static AirOptionEntry fd_options[]  = {
    AF_OPTION( "in", 'i', 0, AF_OPT_TYPE_STRING, &fd_inspec,
               "Input specifier", "inspec" ),
    AF_OPTION( "out", 'o', 0, AF_OPT_TYPE_STRING, &fd_outspec,
               "Output specifier", "outspec" ),
    AF_OPTION( "nextdir", 'n', 0, AF_OPT_TYPE_STRING, &fd_nextdir,
               "Directory to move good input to (or 'delete')", "dir" ),
    AF_OPTION( "faildir", 'x', 0, AF_OPT_TYPE_STRING, &fd_faildir,
               "Directory to move failed input to (or 'delete')", "dir" ),
    AF_OPTION( "extension", 'e', 0, AF_OPT_TYPE_STRING, &fd_outext,
               "Extension to use for output files", "extension" ),
    AF_OPTION( "poll", 'p', 0, AF_OPT_TYPE_INT, &fd_poll_delay,
               "Polling delay in seconds", "sec" ),
    AF_OPTION( "lock", 'k', 0, AF_OPT_TYPE_NONE, &fd_lock,
               "Use exclusive .lock files for concurrency", NULL ),
    AF_OPTION( "pidfile", 'P', 0, AF_OPT_TYPE_STRING, &fd_pidfile,
               "A filename to write the daemon's pid to", NULL ),
    AF_OPTION( "no-daemon", (char)0, 0, AF_OPT_TYPE_NONE, &fd_nodaemon,
               "do not daemonize", NULL ),
    AF_OPTION_END
};


static void
parse_options(
    int   *argc,
    char **argv[])
{
    AirOptionCtx *aoctx = NULL;

    aoctx = air_option_context_new("", argc, argv, fd_options);
    logc_add_option_group(aoctx, "filedaemon", VERSION);

    air_option_context_set_help_enabled(aoctx);

    air_option_context_parse(aoctx);
}


typedef struct _fd_write_data {
    GIOChannel  *infile;
    gchar       *buf;
} fd_write_data_t;

typedef struct _fd_read_data {
    GIOChannel  *outfile;
    gchar       *buf;
    GMainLoop   *loop;
} fd_read_data_t;

static gboolean
write_to_child(
    GIOChannel    *child_stdin,
    GIOCondition   condition,
    gpointer       data)
{
    GIOStatus     ret;
    GError       *err    = NULL;
    gsize         bytes_read, bytes_written;
    GIOChannel   *infile =  ((fd_write_data_t *)data)->infile;
    gchar        *buf    =  ((fd_write_data_t *)data)->buf;
    /** number of attempts for temporarily busy resource */
    unsigned char stillRetry = 10;

    static int    br         = 0;

    if (condition & G_IO_HUP) {
        g_critical("Write end of pipe died!");
    }
    if (condition & G_IO_ERR) {
        g_critical("Error writing to child process");
    }

    while (stillRetry) {
        ret = g_io_channel_read_chars(infile, buf, fd_bufsize,
                                      &bytes_read,
                                      &err);
        switch (ret) {
          case G_IO_STATUS_ERROR:
            g_critical("Error reading: %s", err->message);
            break;
          case G_IO_STATUS_EOF:
            g_debug("eof from infile");
            g_io_channel_shutdown(child_stdin, TRUE, &err);
            return FALSE;
          case G_IO_STATUS_AGAIN:
            g_debug("resource temporarily busy");
            stillRetry--;
            continue;
          default:
            break;
        }
        break;
    }

    br += bytes_read;

    /* g_debug("Read %u bytes from input file (total: %lu).", bytes_read,
     * br);*/

    ret = g_io_channel_write_chars(child_stdin, buf, bytes_read, &bytes_written,
                                   &err);

    if (ret == G_IO_STATUS_ERROR) {
        g_critical("Error writing: %s", err->message);
    }

    /* g_debug("Wrote %u bytes to child.", bytes_written); */
    return TRUE;
}


static gboolean
read_from_child(
    GIOChannel    *child_stdout,
    GIOCondition   condition,
    gpointer       data)
{
    GIOStatus     ret;
    GError       *err     = NULL;
    gsize         bytes_read, bytes_written;
    GIOChannel   *outfile = ((fd_read_data_t *)data)->outfile;
    gchar        *buf     =  ((fd_read_data_t *)data)->buf;
    GMainLoop    *loop    = ((fd_read_data_t *)data)->loop;
    /** number of times to retry a temporary busy read */
    unsigned char stillRetry = 10;

    if (condition & G_IO_HUP) {
        g_critical("Read end of pipe died!");
    }
    if (condition & G_IO_ERR) {
        g_critical("Error reading from child process");
    }

    while (stillRetry) {
        ret = g_io_channel_read_chars(child_stdout, buf, fd_bufsize,
                                      &bytes_read, &err);
        switch (ret) {
          case G_IO_STATUS_ERROR:
            g_critical("Error reading: %s", err->message);
            break;
          case G_IO_STATUS_EOF:
            g_debug("EOF from child (read %u)", (unsigned int)bytes_read);
            g_io_channel_shutdown(outfile, TRUE, &err);
            g_io_channel_shutdown(child_stdout, TRUE, &err);
            g_main_loop_quit(loop);
            return FALSE;
          case G_IO_STATUS_AGAIN:
            g_debug("resource temporarily busy");
            stillRetry--;
            continue;
          default:
            break;
        }
        break;
    }

    /* g_debug("Read %u bytes from child", bytes_read); */

    ret = g_io_channel_write_chars(outfile, buf, bytes_read, &bytes_written,
                                   &err);
    if (ret == G_IO_STATUS_ERROR) {
        g_critical("Error writing: %s", err->message);
    }

    /* g_debug("Wrote %u bytes to output file.", bytes_written); */
    return TRUE;
}


static void
on_child_exit(
    GPid       child_pid,
    gint       status,
    gpointer   data)
{
    char *infile  = (char *)data;
    char *destdir = NULL;

    g_message("pid %lu exited with status %d", (gulong)child_pid, status);

#ifdef G_OS_UNIX
    if (WIFEXITED(status)) {
        if (WEXITSTATUS(status) == EXIT_SUCCESS) {
            g_debug("pid %lu returned success", (gulong)child_pid);
            if (g_ascii_strcasecmp(fd_nextdir, "delete")) {
                destdir = fd_nextdir;
            }
        } else {
            g_warning("pid %lu returned error status %d", (gulong)child_pid,
                      WEXITSTATUS(status));
            if (g_ascii_strcasecmp(fd_faildir, "delete")) {
                destdir = fd_faildir;
            }
        }
    } else if (WIFSIGNALED(status)) {
        g_critical("pid %lu terminated with signal %d\n",
                   (gulong)child_pid, WTERMSIG(status));
        if (g_ascii_strcasecmp(fd_faildir, "delete")) {
            destdir = fd_faildir;
        }
    } else {
        g_critical("pid %lu terminated", (gulong)child_pid);
    }
#endif /* G_OS_UNIX */

    /* Do move or delete */
    if (destdir) {
        if (*destdir) {
            GString *destpath = g_string_new(NULL);
            char    *dbase    = NULL;
            /* Calculate move destination path */
            dbase = g_path_get_basename(infile);
            g_string_printf(destpath, "%s/%s", destdir, dbase);
            if (dbase) {free(dbase);}
            /* Do link */
            g_message("moving %s -> %s", infile, destpath->str);
            if (link(infile, destpath->str) < 0) {
                g_critical(
                    "error moving input file to destination directory: %s",
                    strerror(errno));
            }
            g_string_free(destpath, TRUE);
        }
    }

    /* Do delete */
    if (unlink(infile) < 0) {
        g_critical("error deleting input file");
    }

    g_spawn_close_pid(child_pid);
}


static void
fd_lock_file(
    char  *filename)
{
    GString *lockpath = NULL;
    int      fd       = -1;

    lockpath = g_string_new(NULL);

    g_string_printf(lockpath, "%s.lock", filename);

    /* Attempt lock */
    fd = open(lockpath->str, O_WRONLY | O_CREAT | O_EXCL, 0664);
    if (fd < 0) {
        goto done;
    }
    close(fd);

    /* Verify existence */
    if (!g_file_test(filename, G_FILE_TEST_IS_REGULAR)) {
        /* file not here; unlock it */
        if (lockpath->str) {unlink(lockpath->str);}
    }

  done:
    if (lockpath) {g_string_free(lockpath, TRUE);}
}


static void
fd_unlock_file(
    char  *filename)
{
    GString *lockpath = NULL;

    lockpath = g_string_new(NULL);

    g_string_printf(lockpath, "%s.lock", filename);

    /* Unlock file */
    if (lockpath) {
        unlink(lockpath->str);
        if (lockpath) {g_string_free(lockpath, TRUE);}
    }
}


static gboolean
daemonize(
    void)
{
    /* fork */
    if (fork()) {exit(0);}

    /* dissociate from controlling terminal */
    if (setsid() < 0) {
        g_critical("setsid() failed: %s", strerror(errno));
        return FALSE;
    }

    /* redirect stdio */
    freopen("/dev/null", "r", stdin);
    freopen("/dev/null", "w", stdout);
    freopen("/dev/null", "w", stderr);

    fd_pid = getpid();
    if (fd_pidfile) {
        FILE *pidfile = fopen(fd_pidfile, "w");
        if (!pidfile) {
            g_critical("could not write pidfile");
            goto end;
        }
        fprintf(pidfile, "%d\n", fd_pid);
        fclose(pidfile);
    }
  end:

    return TRUE;
}


int
main(
    int    argc,
    char  *argv[])
{
    glob_t     gbuf;
    int        grc, i;
    GError    *err = NULL;

    GMainLoop *loop;

    GPtrArray *child_args     = NULL;

    /* parse options */
    parse_options(&argc, &argv);

    /* set up logging */
    if (!logc_setup(&err)) {
        air_opterr("%s", err->message);
    }

    if (fd_nextdir == NULL) {
        air_opterr("The --nextdir switch is required");
    }

    child_args = g_ptr_array_sized_new(64);
    for (i = 1; i < argc; i++) {
        /* Double dash indicates end of filedaemon's arguments */
        if (!strncmp(argv[i], "--", strlen(argv[i])) ) {
            continue;
        }
        g_ptr_array_add(child_args, g_strdup(argv[i]));
    }
    g_ptr_array_add(child_args, NULL);

    if (child_args->len > 1) {
        if (fd_faildir == NULL) {
            air_opterr("The --faildir switch is required");
        }
    }

    loop = g_main_loop_new(NULL, FALSE);

    /* We need an input glob */
    if (!fd_inspec) {
        air_opterr("Input glob must be specified");
    }

    /* If an output destination is provided, make sure it's a directory */
    if (fd_outspec && !g_file_test(fd_outspec, G_FILE_TEST_IS_DIR )) {
        air_opterr("Output is not a directory");
    }

    /* Options check out; daemonize */
    if (!fd_nodaemon) {
        if (!daemonize()) {
            goto end;
        }
    }

    while (1) {
        /* Evaluate glob expression */
        grc = glob(fd_inspec, 0, NULL, &gbuf);

        if (grc == GLOB_NOSPACE) {
            g_error("Out of memory: glob allocation failure");
        }
#ifdef GLOB_NOMATCH
        /* HaX0riffic! Simulate behavior without NOMATCH where we have it. */
        else if (grc == GLOB_NOMATCH) {
            gbuf.gl_pathc = 0;
            gbuf.gl_pathv = NULL;
        }
#endif /* ifdef GLOB_NOMATCH */

        /* Iterate over glob paths, enqueueing. */
        for (i = 0; i < (int)gbuf.gl_pathc; i++) {
            char         **child_envp            = {NULL};
            GError        *child_err             = NULL;

            GString       *filename_in           = NULL;
            GString       *filename_out          = NULL;
            GString       *filename_lock         = NULL;

            GIOChannel    *file_in               = NULL;
            GIOChannel    *file_out              = NULL;

            GIOChannel    *child_stdin           = NULL;
            gint           child_stdin_fd        = -1;

            GIOChannel    *child_stdout          = NULL;
            gint           child_stdout_fd       = -1;

            GPid           child_pid;
            int            len;

            fd_read_data_t read_data;
            fd_write_data_t write_data;

            filename_in = g_string_new(gbuf.gl_pathv[i]);

            /* Skip non-regular files */
            if (!g_file_test(filename_in->str, G_FILE_TEST_IS_REGULAR) ) {
                g_string_free(filename_in, TRUE);
                continue;
            }

            /* Skip lockfiles */
            if (!strcmp(".lock", filename_in->str
                        + strlen(filename_in->str) - 5))
            {
                g_string_free(filename_in, TRUE);
                continue;
            }

            /* Generate lock path */
            if (!filename_lock) {filename_lock = g_string_new(NULL);}
            g_string_printf(filename_lock, "%s.lock", filename_in->str);

            /* Skip files locked at queue time */
            if (g_file_test(filename_lock->str, G_FILE_TEST_IS_REGULAR)) {
                g_debug("file %s is locked", filename_in->str);
                g_string_free(filename_in, TRUE);
                g_string_free(filename_lock, TRUE);
                continue;
            }

            if (child_args->len == 1) {
                /* Do move or delete */
                GString *destpath = g_string_new(NULL);
                char    *dbase    = NULL;
                /* Calculate move destination path */
                dbase = g_path_get_basename(filename_in->str);

                g_string_printf(destpath, "%s/%s", fd_nextdir, dbase);
                if (dbase) {free(dbase);}
                /* Do link */
                g_message("moving %s -> %s", filename_in->str, fd_nextdir);
                if (link(filename_in->str, destpath->str) < 0) {
                    g_critical(
                        "error moving input file to destination directory: %s",
                        strerror(errno));
                }
                /* Do delete */
                if (unlink(filename_in->str) < 0) {
                    g_critical("error deleting input file");
                }
                g_string_free(destpath, TRUE);
                g_string_free(filename_in, TRUE);
                g_string_free(filename_lock, TRUE);
                continue;
            }

            if (fd_lock) {
                fd_lock_file(filename_in->str);
            }

            file_in = g_io_channel_new_file(filename_in->str, "r", &err);

            if (file_in == NULL) {
                g_critical("Cannot open input file!");
            }

            g_io_channel_set_encoding(file_in, NULL, &err);

            if (err) {
                g_critical("error setting input encoding!");
            }

            g_io_channel_set_buffer_size(file_in, fd_bufsize);

            filename_out = g_string_new(NULL);

            if (fd_outspec == NULL) {
                g_string_printf(filename_out, "%s", gbuf.gl_pathv[i]);
            } else {
                g_string_printf(filename_out, "%s/%s", fd_outspec,
                                gbuf.gl_pathv[i]);
            }

            len  = filename_out->len;

            if (g_strrstr(filename_out->str, ".")) {
                while (len-- > 0
                       && !g_str_has_suffix(filename_out->str, ".") )
                {
                    g_string_set_size(filename_out, filename_out->len - 1);
                }
                g_string_set_size(filename_out, filename_out->len - 1);
            }
            if (fd_outext) {
                g_string_append_printf(filename_out, ".%s", fd_outext);
            } else {
                g_string_append(filename_out, ".out");
            }

            g_message("%d: %s -> %s", i, filename_in->str, filename_out->str);

            file_out = g_io_channel_new_file(filename_out->str, "w", &err);

            if (file_out == NULL) {
                g_error("Cannot open output file!");
            }

            g_io_channel_set_encoding(file_out, NULL, &err);

            if (err) {
                g_error("error setting output encoding!");
            }

            g_io_channel_set_buffer_size(file_out, fd_bufsize);

            if (!g_spawn_async_with_pipes(".",
                                          (gchar **)child_args->pdata,
                                          child_envp,
                                          G_SPAWN_SEARCH_PATH |
                                          G_SPAWN_DO_NOT_REAP_CHILD,
                                          NULL,
                                          NULL,
                                          &child_pid,
                                          &child_stdin_fd,
                                          &child_stdout_fd,
                                          NULL,
                                          &child_err))
            {
                g_error("error spawning process: %s",
                        (child_err && child_err->message ? child_err->
                         message : "unknown error"));
            }
            g_debug("spawned process %d", i);

            /* Watch for process exit status */
            g_child_watch_add(child_pid, on_child_exit, filename_in->str);

            child_stdin = g_io_channel_unix_new(child_stdin_fd);

            if (child_stdin == NULL) {
                g_error("Cannot open child stdin!");
            }

            g_io_channel_set_encoding(child_stdin, NULL, &err);

            if (err) {
                g_error("error setting child stdin encoding!");
            }

            g_io_channel_set_buffer_size(child_stdin, fd_bufsize);

            child_stdout = g_io_channel_unix_new(child_stdout_fd);

            if (child_stdout == NULL) {
                g_error("Cannot open child stdout!");
            }

            g_io_channel_set_encoding(child_stdout, NULL, &err);

            if (err) {
                g_error("error setting child stdout encoding!");
            }

            g_io_channel_set_buffer_size(child_stdout, fd_bufsize);

            write_data.infile = file_in;
            write_data.buf    = g_malloc(g_io_channel_get_buffer_size(file_in));

            if (write_data.buf == NULL) {
                g_error("error allocating file_in buffer");
            }

            if (!g_io_add_watch(child_stdin,  G_IO_OUT | G_IO_PRI | G_IO_HUP |
                                G_IO_ERR, write_to_child, &write_data))
            {
                g_error("Cannot add watch on GIOChannel!");
            }

            read_data.outfile = file_out;
            read_data.loop    = loop;
            read_data.buf     = g_malloc(g_io_channel_get_buffer_size(
                                             file_out));

            if (write_data.buf == NULL) {
                g_error("error allocating file_in buffer");
            }

            if (!g_io_add_watch(child_stdout, G_IO_IN | G_IO_PRI | G_IO_HUP |
                                G_IO_ERR, read_from_child, &read_data))
            {
                g_error("Cannot add watch on GIOChannel!");
            }

            g_main_loop_run(loop);

            if (fd_lock) {
                fd_unlock_file(filename_in->str);
            }

            if (read_data.buf) {
                g_free(read_data.buf);
            }
            if (write_data.buf) {
                g_free(write_data.buf);
            }

            g_string_free(filename_in, TRUE);
            g_string_free(filename_out, TRUE);
            g_string_free(filename_lock, TRUE);
        }
        sleep(fd_poll_delay);
    }

  end:

    return 0;
}
